scanpy/preprocessing/combat.py

```diff
@@ -0,0 +1,161 @@
+import numpy as np
+from scipy.sparse import issparse
+import pandas as pd
+import sys
+from numpy import linalg as la
+import patsy
+
+def design_mat(mod, batch_levels):
+    # require levels to make sure they are in the same order as we use in the
+    # rest of the script.
+    design = patsy.dmatrix("~ 0 + C(batch, levels=%s)" % str(batch_levels),
+                                                  mod, return_type="dataframe")
+
+    mod = mod.drop(["batch"], axis=1)
+    sys.stderr.write("found %i batches\n" % design.shape[1])
+    other_cols = [c for i, c in enumerate(mod.columns)]
+    factor_matrix = mod[other_cols]
+    design = pd.concat((design, factor_matrix), axis=1)
+    sys.stderr.write("found %i categorical variables:" % len(other_cols))
+    sys.stderr.write("\t" + ", ".join(other_cols) + '\n')
+    return design
+
+def combat(adata, key = 'batch'):
+    """Correct for batch effects in a dataset. Expects normalised, log-transformed data.
+    This is 
+
+    Parameters
+    ----------
+    adata : AnnData object 
+    key: str
+        key to a categorical annotation from adata.obs that will be used for batch effect removal
+    copy: bool
+        wether to update the adata object or to copy it
+
+    Returns
+    -------
+    Depending on the value of copy, either returns an updated AnnData object or modifies the passed one
+    """
+    
+    # only works on dense matrices so far
+    if issparse(adata.X): 
+        X = adata.X.A.T
+    else: 
+        X = adata.X.T
+    data = pd.DataFrame(data = X, index = adata.var_names,
```

Comment:
Please, no white spaces when passing keyword parameters: https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements

Reply:
thanks!